src/lib/Repository/Helper/NameSchemaService.php

```diff
@@ -6,27 +6,190 @@
+            $eventDispatcher,
+            $settings
+        );
+        $this->contentTypeHandler = $contentTypeHandler;
+        $this->contentTypeDomainMapper = $contentTypeDomainMapper;
+    }
+
     public function resolveUrlAliasSchema(Content $content, ContentType $contentType = null): array
     {
         $contentType = $contentType ?? $content->getContentType();
 
-        return $this->resolve(
+        return $this->resolveNameSchema(
             empty($contentType->urlAliasSchema) ? $contentType->nameSchema : $contentType->urlAliasSchema,
             $contentType,
             $content->fields,
             $content->versionInfo->languageCodes
         );
     }
+
+    public function resolveContentNameSchema(
+        Content $content,
+        array $fieldMap = [],
+        array $languageCodes = [],
+        ContentType $contentType = null
+    ): array {
+        $contentType ??= $content->getContentType();
+
+        $languageCodes = $languageCodes ?: $content->versionInfo->languageCodes;
+
+        return $this->resolveNameSchema(
+            $contentType->nameSchema,
+            $contentType,
+            $this->mergeFieldMap(
+                $content,
+                $fieldMap,
+                $languageCodes
+            ),
+            $languageCodes
+        );
+    }
+
+    public function resolveNameSchema(
+        string $nameSchema,
+        ContentType $contentType,
+        array $fieldMap,
+        array $languageCodes
+    ): array {
+        [$filteredNameSchema, $groupLookupTable] = $this->filterNameSchema($nameSchema);
+        $tokens = $this->extractTokens($filteredNameSchema);
+        $schemaIdentifiers = $this->getIdentifiers($nameSchema);
+
+        $names = [];
+
+        foreach ($languageCodes as $languageCode) {
+            // Fetch titles for language code
+            $titles = $this->getFieldTitles($schemaIdentifiers, $contentType, $fieldMap, $languageCode);
+            $name = $filteredNameSchema;
+
+            // Replace tokens with real values
+            foreach ($tokens as $token) {
+                $string = $this->resolveToken($token, $titles, $groupLookupTable);
+                $name = str_replace($token, $string, $name);
+            }
+            $name = $this->validateNameLength($name);
+
+            $names[$languageCode] = $name;
+        }
+
+        return $names;
+    }
+
+    protected function mergeFieldMap(Content $content, array $fieldMap, array $languageCodes)
+    {
+        if (empty($fieldMap)) {
+            return $content->fields;
+        }
+
+        $mergedFieldMap = [];
+
+        foreach ($content->fields as $fieldIdentifier => $fieldLanguageMap) {
+            foreach ($languageCodes as $languageCode) {
+                $mergedFieldMap[$fieldIdentifier][$languageCode] = isset($fieldMap[$fieldIdentifier][$languageCode])
+                    ? $fieldMap[$fieldIdentifier][$languageCode]
+                    : $fieldLanguageMap[$languageCode];
```

Comment:
Can be replaced with:
```suggestion
                $mergedFieldMap[$fieldIdentifier][$languageCode] = $fieldMap[$fieldIdentifier][$languageCode] ?? $fieldLanguageMap[$languageCode];
```